### PR TITLE
QTPFSv2

### PR DESCRIPTION
**Pull request by marcushutchings**
QTPFS updated to become a viable alternative to HAPFS for large Recoil game projects.

Main issues with the original QTPFS:
* Consumed way too much memory, loading the map, Epic, in ZK or BAR could crash machines. BAR used 16.3GB when it reached the deployment stage of a match.
* Some maps, like flats and green, would be a slideshow for the first minute or so because the system was bogged down trying to process over 5 million map damage updates, recorded on frame 0.
* Sim speed would drop to x0.1 on `fightertest 200 armbull armbull` mainly because map damage updates were too expensive.
* Didn't support unsync pathing requests.

Why bother to update QTPFS over using HAPFS?
* QTPFS provides paths that are not resitricted to 8-way movement, like HAPFS.
* HAPFS provides a raw move hack to make movement look better, but it then ignores the slow-down affect of terrain causing units to potentially travel routes that take longer to reach the goal. *
* HAPFS raw move cannot handle lava.
* QTPFS paths don't need raw move, and so can handle pathing around lava while also providing better looking paths. *
* QTPFS paths are easier for fast vehicles to follow without them having to slalom.
* QTPFS debug view better shows off the accessible areas of the map.
* QTPFS can resolve paths for groups of units faster than HAPFS.
* HAPFS has situations where it may fail to path correctly due to how it works. QTPFS shouldn't have this issue.
* QTPFS can reroute units automatically if the current path is no longer viable.
* QTPFS offers the possibility (and it's just possibility at this stage) or allowing for true bridges, or for exit-only areas in factories.
* In BAR with HAPFS units can sometimes get stuck in their factories. This is partly due to imprecision on pathing rules. QTPFS should allow units to escape in such situations.

_Note: * QTPFSv2 now provides raw move as well. During testing, line move commands accross open fields look too good to the player with raw move that even QTPFS's smoothed move cannot compete. As such, QTPFSv2 will attempt a raw move; however, unlike HAPFS, some provision to mitigate lava areas has been provided with `pfRawMoveSpeedThreshold`; this, may not be enough, but we will try it for now and are open to better suggestions on how to detect lava areas during raw path check._

Major changes in QTPFSv2:
* The update system now uses a dirty quad table, so multiple updates in the same area won't trigger unneccesary path data updates. On flats and green, the initial set of updates needed dropped from 5 million+ to around 5,000.
* The update system has been rebalanced to keep up with map updates while minimizing the cost of doing so. It scales up work if the number of updates starts to get larger. Update have about the same cost as it is for HAPFS, but responds to map updates around 50% faster.
* The implementation is now hybrid Quad-Tree. Instead of one root-node, there are now multiple root nodes in a uniform grid across the map. This is primarily to ensure that all quad sub divide into powers of 2, which affords a series of assumptions that allow for optimizations.
* This multi-root approach also fixes a problem with the original QTPFS where long maps would subdivide into long and sub-optimal rectangler shapes - like a 1x16 squares for example.
* The update system is now multi-threaded.
* Pathing requests are now fully multi-threaded. Previously the requests were single threaded, but in a separate thread from the main one.
* If a pathing request goal is an unaccesible quad, then the system will try to find the nearest accessible node to search instead. For example if move request happens to put the goal on a part of a slope a unit can't move on (or a wreck) then the system originally had to search the whole map - whereas now, it paths to the nearest open node. So what would take 23ms now takes <1ms.
* Pathing requests are issued much sooner. Previously, it could take several seconds for a large group to all receive their paths, now it is within 1-3 frames.
* Pathing requests will now do a raw-move pass first to see if the unit can walk in a straight line to the goal.
* The node-related data structures have been reworked to require significantly less memory than before. Also, excess data links have been removed.
* QTPFS could group related pathing requests made in the same frame to reduce load. Now, all pathing requests can be grouped with any active pathing request. So a request result processed in a previous frame can be used as the answer for one in the current frame if they are travelling from/to the same quads.
* Support for unsynced pathing requests added.
* No longer saves/loads pathing data to/from disk: with the performance improvements it is faster to recalculate the pathing data on load than to use data saved to disk.
* Introduces the EnTT ECS system to the Recoil Engine.

Performance of QTPFSv2:
* In the example where BAR loaded Epic and needed 16.3GB (in total for the whole process, not just mapping), this version needed about 7.1GB. For reference, HAPFS needed 5.2GB. HAPFS has an advantage in this regard because all map data is stored effectively as though it is the max-res HAPFS level, whereas QTPFS has to translate that into a quad-tree structure.
* In flats and green, the experience is smooth, and all initial frame 0 map damage is processed in about 20 seconds.
* Sim speed with `fightertest 200 armbull armbull` is now typically between x1.7 and x2.2. HAPFS is typically between x1.7 and x1.9
* Very long paths are where QTPFS can be slower than HAPFS - HAPFS can delay resolved full-detail paths whereas QTPFS can to build the full path.

What is not included:
* There is an old (and probably now considered deprecated) interface for specifying an Extra Cost Map, which could be used to add/subtract weight for pathing. It was originally for the old AI interface, but doesn't seem to be used by our current AIs. givne that it would increase memory usage and reduce performance, I've elected to not support it at this time. This can be revisited it new information comes to light that may suggest otherwise. HAPFS has support for it, but the original QTPFS doesn't.

Limitations of QTPFS:
* Move cost calculations are done on area and are not directional (unlike HAPFS), so slopes will have a fixed move cost regardless of whether the unit would be going up or down. This means units will have a stronger preference to avoid sloping terrain.
* Very rarely groups of units can find issue with very specific tight corners of inaccessible terrain, where they struggle to follow around it as smoothly as other corners. This happens because the engine only samples the unit's mid-point for terrain collisions.

Other Movement-related Changes: *Note: these benefits apply regardless of which path finder you use*
* Fixed: units could sometimes get stuck as they spin/circle around a waypoint when a waypoint is too close to some small/tiny obstacles.
* Fixed: units could sometimes consume waypoints that were well ahead of their current position, which increased processing overhead and reduced their ability to navigate obstacles.
* Fixed: units were often setting their current and next waypoints to the same waypoint.
* Units repath when they are starting to struggle to make progression on their path. Previously they would request a new path upon touching blocking terrain or a building, which may not be necessary and caused a not insignificant impact on CPU usage.
* Reduced the amount of corner-cutting units try to do when following a path. This should reduce the difficulty some unit types experience when trying to move through narrow corridoors, between buildings in bases.
* Reduced the difficulty units face when trying to rejoin a path after being routed away due to a collision with a building.
* Units no longer wait for a path before starting to move, which reduces visible command latency.
* Fixes https://github.com/beyond-all-reason/spring/issues/374

### CODE REVIEW

**Comment by sprunk** [comment]
What happens if the entire map is modified? In particular what if I do `Spring.AdjustHeightMap(0, 0, Game.mapSizeX, Game.mapSizeZ, 123)` at runtime (when there are some units around, or if I do it at a frequency of once every N frames for a duration of M frames where N is some single-digit number and M is some two-three digit number).

**Comment by marcushutchings** [comment]
Very good question. If every sector is updated, then the path updater has to work hard to try and update the map fast enough so that pathing solutions are corrected in a reasonable time-frame (though 'reasonable' may mean something different to different games so the intensitity of the ramping up can be configured.) In my testing, at full load, the updater was using between 20-30% CPU for the first few seconds (and then it begins to drop off), which is fine if you are at the start of a match because nothing else needs enough CPU to cause trouble. However, you wouldn't want that kind of load late-game mid-battle.

This is the same for HAPFS as well. That was testing with BAR using 60 movetypes; so if you only used 12 for example then the CPU usage should drop to 4-6% for example. You also have to consider that the bigger the map, the longer it will take to update the path data and correct unit paths. We're talking tens of seconds, so that may not be good for games in general to damage the whole map at once.

Having lava levels rising and falling is something BAR really wants to do, but we need to be realistic about what is achievable. So if you could keep the map damage to 10% of the map (though various techniques, designs and optimizations) for example, that would go a long way to helping.

**Comment by marcushutchings** [comment]
Ultimately it is something we can experiment with. We have some configurability options so allow adjustments to get the balance right.

**Comment by 6AKU** [comment]
Any ETA when it's gonna be available for testing?

**Comment by marcushutchings** [comment]
> Any ETA when it's gonna be available for testing?

I'll try to make a version available for testing by the end of this weekend - as long as you can take an engine build and use it. I aim to have the code fully tidied up by the end of the month.

File: doc/changelog.txt [hunk 1]
```diff
@@ -89,6 +89,29 @@ Sim:
...
+   or whether a new path should be requested. Defaults to 60 frames (2 seconds.) Adjust to find the
+   right balance between how quickly units can get unstuck and how much CPU power is used. Smaller
+   intervals increase the chance of slow units triggering unnecessary re-pathing requests, but
+   reduces the chance that players may believe a unit is getting stuck and is not handling itself
+   well.
+ - Added mod option, pfRepathMaxRateInFrames, which tells GroundMoveType the minimum amount of
+   frames that must pass before a unit is allowed to request a new path. By default, it is 150
+   (5 seconds.) This is mostly for rate limiting and prevent excessive CPU wastage, because
+   processing path requests are not cheap.
+
```
Comment (sprunk) [inline doc/changelog.txt:110]: Seems some modrule changes are missing?
```suggestion
 - removed modrules: `system.pfForce[Update]SingleThreaded`. MT has shown itself stable
 - removed modrule: `system.pathFinderUpdateRate`. The new approach to perf tradeoffs is <FIXME>
 
```

```

File: rts/Rendering/HAPFSPathDrawer.cpp [hunk 2]
```diff
@@ -299,13 +299,13 @@ void HAPFSPathDrawer::Draw() const {
 
 		glBegin(GL_LINE_STRIP);
 
-			// draw low-res segments of <path> (green)
+			// draw low-res segments of <path> (blue)
```
[the remark this answers is not in the document]
Comment (marcushutchings) [reply rts/Rendering/HAPFSPathDrawer.cpp:302]: Fair point. I will try to remember in future.

File: doc/changelog.txt [hunk 3]
```diff
@@ -89,6 +89,29 @@ Sim:
    will still always target enemies and never allies, so avoid using it outside of manually-triggered contexts
  - nanoturret (immobile builder) build-range now only needs to reach the edge of the buildee's radius
    instead of its center. Mobile builders already worked this way.
+ - damage and armor can also be negative again (like in 0.76b1, so that the target is healed), but
+   keep in mind weapons will still always target enemies and never allies, so avoid using it
+   outside of manually-triggered contexts
+ - Renamed TKPFS to HAPFS. It is HAPFS will modifications and is ready now to take over. It was
+   kept separate initially in case major issues were found, but it is now suitably stable to
+   become the official version of HAPFS (as it was always intended.)
+ - Major rework of QTPFS, so many changes that this diserves its own section.
```
[the remark this answers is not in the document]
Comment (marcushutchings) [reply doc/changelog.txt:98]: Like I said at the top of the page. The changelog is the one thing I'm still sorting out.

File: doc/changelog.txt [hunk 4]
```diff
@@ -89,6 +89,29 @@ Sim:
    will still always target enemies and never allies, so avoid using it outside of manually-triggered contexts
  - nanoturret (immobile builder) build-range now only needs to reach the edge of the buildee's radius
    instead of its center. Mobile builders already worked this way.
+ - damage and armor can also be negative again (like in 0.76b1, so that the target is healed), but
+   keep in mind weapons will still always target enemies and never allies, so avoid using it
+   outside of manually-triggered contexts
```
[the remark this answers is not in the document]
Comment (marcushutchings) [reply doc/changelog.txt]: Removed.

File: doc/changelog.txt [hunk 5]
```diff
@@ -89,6 +89,29 @@ Sim:
...
+ - Renamed TKPFS to HAPFS. It is HAPFS will modifications and is ready now to take over. It was
```
[the remark this answers is not in the document]
Comment (marcushutchings) [reply doc/changelog.txt]: corrected.

File: doc/changelog.txt [hunk 6]
```diff
@@ -89,6 +89,29 @@ Sim:
...
+   kept separate initially in case major issues were found, but it is now suitably stable to
+   become the official version of HAPFS (as it was always intended.)
+ - Major rework of QTPFS, so many changes that this diserves its own section.
+ - Added mod option, pfRepathDelayInFrames, which tells GroundMoveType how many frames at least
```
[the remark this answers is not in the document]
Comment (marcushutchings) [reply doc/changelog.txt:96]: Done.

File: doc/changelog.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (sprunk) [reply doc/changelog.txt:95]: Ah sorry, only looked at the code and not at the github post. It looks good though and could just be copy-pasted here verbatim I think.

[the remark this answers is not in the document]
Comment (marcushutchings) [reply doc/changelog.txt]: No worries! I'm getting live editing and reviewing! 😸 Anyway, added the change log.

**Comment by marcushutchings** [comment]
Fixes https://github.com/beyond-all-reason/spring/issues/374

**Comment by marcushutchings** [comment]
Final call for feedback before mainline merge.

**Comment by GoogleFrog** [comment]
Sounds good. I was away for those 14 hours but here is some feedback anyway.

> Note: * QTPFSv2 now provides raw move as well. During testing, line move commands accross open fields look too good to the player with raw move that even QTPFS's smoothed move cannot compete. As such, QTPFSv2 will attempt a raw move; however, unlike HAPFS, some provision to mitigate lava areas has been provided with pfRawMoveSpeedThreshold; this, may not be enough, but we will try it for now and are open to better suggestions on how to detect lava areas during raw path check.

I figured this out a while ago. The terminal goal of pathfinding is to do what the player wants, which is not simply to move units from A to B in the shortest time possible. If there are two paths, one of which is 0.1s shorter, while the other is what 99% of people would expect their units to take, then I'd call the latter the correct path. This is especially true over short distances as a common style of micro is to give short-ish orders which really mean "Move in this direction" rather than "Move to this point exactly". Having units walk in straight lines rather than bend due to hard-to-precisely-parse slopes is highly desirable for this form of control.

> Move cost calculations are done on area and are not directional (unlike HAPFS), so slopes will have a fixed move cost regardless of whether the unit would be going up or down. This means units will have a stronger preference to avoid sloping terrain.

I'm not sure how much of limitation this will be. I will have to try it out and get feedback from particularly picky players. I expect it would be an issue over short distances, but if short distances are raw moves then I am unsure.

Otherwise, looks exciting.

**Pull request merged**